### PR TITLE
Refactor training logic

### PR DESCRIPTION
**Pull request by billylanchantin**
Adds:

* `%Training.State{status: :cont | :halt}`

This way the `fun` on `%Training.Callback{}` can have the spec:

* `fun :: State.t() -> State.t()`

instead of wrapping the output in a tuple. It also means we can write code like this:

```elixir
state =
  state
  |> run_callbacks(callbacks, :before_training)
  |> run_training(callbacks, dmat, objective)
  |> run_callbacks(callbacks, :after_training)
```

since now `run_callbacks/3` and `run_training/4` (new) can branch on `state.status`.

I did some other code shuffling while I was in there, but none of it is crucial.

### CODE REVIEW

File: lib/exgboost/training/callback.ex [hunk 1]
```diff
@@ -134,54 +140,34 @@ defmodule EXGBoost.Training.Callback do
...
 
       true ->
-        updated_meta_vars =
-          meta_vars
-          |> put_in([:early_stop, :since_last_improvement], since_last_improvement + 1)
-
-        bst = struct(bst, best_iteration: state.iteration, best_score: cur_criteria_value)
-        {:halt, %{state | meta_vars: updated_meta_vars, booster: bst}}
+        early_stop = Map.update!(early_stop, :since_last_improvement, &(&1 + 1))
+        bst = struct(bst, best_iteration: state.iteration, best_score: score)
```
Comment (billylanchantin) [inline lib/exgboost/training/callback.ex:169]: Is this possibly a bug? It doesn't match the other `cond` branch where:

```elixir
bst =
  bst
  |> struct(best_iteration: state.iteration, best_score: score)
  |> EXGBoost.Booster.set_attr(best_iteration: state.iteration, best_score: score)
```

I haven't had time time to dig into the particulars of what's going on in `Booster` yet, so I wasn't sure.

File: lib/exgboost/training.ex [hunk 2]
```diff
@@ -49,200 +58,137 @@ defmodule EXGBoost.Training do
...
-            init_state: %{learning_rates: learning_rates}
-          }
-          | callbacks
-        ]
-      else
-        callbacks
-      end
+      Enum.map(callbacks ++ defaults, fn %Callback{fun: fun} = callback ->
+        %{callback | fun: fn state -> state |> fun.() |> State.validate!() end}
+      end)
```
Comment (billylanchantin) [inline lib/exgboost/training.ex:67]: This is a tiny bit of overhead, but I think it'll result in better error messages.

File: lib/exgboost/training/callback.ex (hunk 1 above)
Comment (acalejos) [reply lib/exgboost/training/callback.ex:169]: Yeah it should probably either do bot add to struct and update the attirbute or just update the struct. The Booster attributes are held in the actual booster so are actually mutable. The implementation to use attributes exists but I dont think should really be used outside of maybe seeing any attributes from a serialized model that maybe were set elsewhere

Comment (acalejos) [reply lib/exgboost/training/callback.ex:169]: Also Im glad you cleaned up a lot of this code because IDK why I was using `put_in` so much

File: lib/exgboost/training/callback.ex [hunk 3]
```diff
@@ -229,14 +215,14 @@ defmodule EXGBoost.Training.Callback do
...
             monitor_metrics: %{period: period, filter: filter}
-          }
+          },
+          status: :cont
         } = state
       ) do
     if period != 0 and rem(iteration, period) == 0 do
-      metrics = Map.filter(metrics, filter)
-      IO.puts("Iteration #{iteration}: #{inspect(metrics)}")
+      IO.puts("Iteration #{iteration}: #{inspect(Map.filter(metrics, filter))}")
```
Comment (acalejos) [inline lib/exgboost/training/callback.ex:223]: Let me know if you have any opinions about improving this logging. It's pretty bare-bones right now. Not necessarily for this PR though

File: lib/exgboost/training/state.ex [hunk 4]
```diff
@@ -3,9 +3,18 @@ defmodule EXGBoost.Training.State do
   @enforce_keys [:booster]
   defstruct [
     :booster,
-    meta_vars: %{},
     iteration: 0,
     max_iteration: -1,
-    metrics: %{}
+    meta_vars: %{},
+    metrics: %{},
```
Comment (acalejos) [inline lib/exgboost/training/state.ex:9]: Im trying to remember why I didnt just use the `meta_vars` to hold the metrics in...

File: lib/exgboost/training/callback.ex [hunk 5]
```diff
@@ -46,30 +46,35 @@ defmodule EXGBoost.Training.Callback do
...
-      when event in [:before_training, :after_training, :before_iteration, :after_iteration] and
-             is_atom(name) do
+  @spec new(event :: event(), fun :: fun(), name :: atom(), init_state :: any()) :: Callback.t()
+  def new(event, fun, name, init_state \\ %{})
+      when event in @valid_events and is_function(fun, 1) and is_atom(name) and not is_nil(name) do
     %__MODULE__{event: event, fun: fun, name: name, init_state: init_state}
+    |> validate!()
+  end
+
+  def validate!(%__MODULE__{} = callback) do
```
Comment (acalejos) [inline lib/exgboost/training/callback.ex:64]: My only hangup here is that the only reason we need a name is to look up associated `meta_vars`, so there should still be the ability to create a callback that has no state and no name

Comment (billylanchantin) [reply lib/exgboost/training/callback.ex:64]: Oh I totally forgot about this part. Sorry I meant to call it out.

My thinking was this: we can't use `name` as the key in `meta_vars` unless all names are unique. Multiple callbacks with `name: nil` break that. Since I couldn't find an instance where `name: nil` was actually used, I figured I'd just make it required.

If we _do_ want to allow `name: nil`, then we need to alter the creation of `meta_vars` to account for duplicates. Come to think of it, we need to do that anyway.

File: lib/exgboost/training/callback.ex (hunk 3 above)
Comment (billylanchantin) [reply lib/exgboost/training/callback.ex:223]: I think bare-bones is fine. And yeah, not for this PR.

However, an example in the docs that overrides this callback to, say, use `Logger` might be quite instructive.

(Also this was such I nit-picky change on my part, lol. The code before was completely fine.)

File: lib/exgboost/training/callback.ex (hunk 5 above)
Comment (acalejos) [reply lib/exgboost/training/callback.ex:64]: My thought was that you just check for `nil` name and dont even look at meta vars if its `nil` and instead just run the func. And then the `name`s would have to be unique still

Comment (billylanchantin) [reply lib/exgboost/training/callback.ex:64]: I addressed the duplicate name issue and wrote a test.

[f1977bf](https://github.com/acalejos/exgboost/pull/35/commits/f1977bf9190816eb237505471befb9291b5c675b)

This doesn't address the question about how to handle `name: nil`, but it prevents us clobbering callbacks when we build `meta_vars`. We can easily filter `name: nil` callbacks out or handle them with some other convention.

Thoughts?

Comment (billylanchantin) [reply lib/exgboost/training/callback.ex:64]: FYI, I think the other way to handle this is to do something like:

```elixir
meta_vars =
  callbacks
  |> Enum.sort_by(& &1.name) # for determinism
  |> Enum.with_index()
  |> Map.new(&{&1.name, &1.init_state})
```

That side-steps the name issue altogether. But then if duplicate names are present, it's not possible to provide an error message like "Callback X did the wrong thing.".
